os/services/orchestra/orchestra-rule-special-for-root.c

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (c) 2018, Amber Agriculture
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the Institute nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE INSTITUTE AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE INSTITUTE OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ *
+ */
+/**
+ * \file
+ *         Orchestra: a slotframe dedicated to unicast data transmission to the root.
+ *         See the paper "TSCH for Long Range Low Data Rate Applications", IEEE Access
+ *
+ * \author Atis Elsts <atis.elsts@gmail.com>
+ */
+
+#include "contiki.h"
+#include "orchestra.h"
+
+#include "sys/log.h"
+#define LOG_MODULE "Orchestra"
+#define LOG_LEVEL  LOG_LEVEL_MAC
+
+static struct tsch_slotframe *sf_tx;
+static struct tsch_slotframe *sf_rx;
+static uint8_t is_root_rule_used;
+static uint16_t timeslot_tx;
+
+static void set_self_to_root(uint8_t is_root);
+/*---------------------------------------------------------------------------*/
+static inline uint8_t
+self_is_root(void)
+{
+  /* If this is changed to look at the RPL status instead of tsch_is_coordinator,
+   * multiple roots become possible in a single TSCH network. */
```

Comment:
But ok now reading this comment I'm confused: how does a network with one coordinator but multiple roots work?
What does the time-source tree/dag look like in that case?

Reply:
Hmm ok. So, you have a time-source tree with corresponding routeCost then, independent from the multiple RPL DAGs?